File: backend/app/models/expense.py
from sqlalchemy import Column, DateTime, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database.database import Base


class Expense(Base):
    __tablename__ = "expenses"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(200), nullable=False, index=True)
    amount = Column(Numeric(12, 2), nullable=False)
    expense_date = Column(DateTime(timezone=True), nullable=False, index=True)
    category = Column(String(100), nullable=False, index=True)
    notes = Column(String(1000), nullable=True)
    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False
    )
    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)

    owner = relationship("User", back_populates="expenses")

File: frontend/src/types/expense.ts
export type Expense = {
  id: number
  title: string
  amount: number
  category: string
  notes: string | null
  expense_date: string
  created_at: string
  updated_at: string
}

export type ExpensePayload = {
  title: string
  amount: number
  category: string
  notes: string | null
  expense_date: string
}

export type ExpenseQueryParams = {
  skip?: number
  limit?: number
  category?: string
  start_date?: string
  end_date?: string
  sort_by?: 'created_at' | 'expense_date' | 'amount'
  sort_dir?: 'asc' | 'desc'
  search?: string
}
